### PR TITLE
feat(gossipsub): deprecate `Rpc` in preparation for making it private

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/gossipsub/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 0.46.1 - unreleased
+
+- Deprecate `Rpc` in preparation for removing it from the public API because it is an internal type.
+  See [PR XXXX](https://github.com/libp2p/rust-libp2p/pull/XXXX).
```
Comment (thomaseizinger) [inline protocols/gossipsub/CHANGELOG.md]:
```suggestion
  See [PR 4833](https://github.com/libp2p/rust-libp2p/pull/4833). 
```

File: protocols/gossipsub/src/handler.rs [hunk 2]
```diff
@@ -47,6 +47,8 @@ pub enum HandlerEvent {
     /// any) that were received.
     Message {
         /// The GossipsubRPC message excluding any invalid messages.
+        // `HandlerEvent` shouldn't be public but we can't enforce that.
+        #[allow(private_interfaces)]
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/handler.rs]:
```suggestion
```

File: protocols/gossipsub/src/lib.rs [hunk 3]
```diff
@@ -125,7 +125,11 @@ pub use self::subscription_filter::{
 };
 pub use self::topic::{Hasher, Topic, TopicHash};
 pub use self::transform::{DataTransform, IdentityTransform};
-pub use self::types::{Message, MessageAcceptance, MessageId, RawMessage, Rpc};
+pub use self::types::{Message, MessageAcceptance, MessageId, RawMessage};
+
+#[deprecated(note = "Will be removed from the public API.")] // Once this is done, remove `#[allow(unreachable_pub)]` from it!
+#[allow(private_interfaces)]
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/lib.rs]:
```suggestion
```

File: protocols/gossipsub/src/types.rs [hunk 4]
```diff
@@ -234,7 +234,7 @@ pub struct Rpc {
 impl Rpc {
     /// Converts the GossipsubRPC into its protobuf format.
     // A convenience function to avoid explicitly specifying types.
-    pub fn into_protobuf(self) -> proto::RPC {
+    pub(crate) fn into_protobuf(self) -> proto::RPC {
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/types.rs]:
```suggestion
    pub fn into_protobuf(self) -> proto::RPC {
```

File: protocols/gossipsub/src/lib.rs [hunk 5]
```diff
@@ -125,7 +125,10 @@ pub use self::subscription_filter::{
 };
 pub use self::topic::{Hasher, Topic, TopicHash};
 pub use self::transform::{DataTransform, IdentityTransform};
-pub use self::types::{Message, MessageAcceptance, MessageId, RawMessage, Rpc};
+pub use self::types::{Message, MessageAcceptance, MessageId, RawMessage};
+
+#[deprecated(note = "Will be removed from the public API.")] // Once this is done, remove `#[allow(unreachable_pub)]` from it!
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/lib.rs]:
```suggestion
#[deprecated(note = "Will be removed from the public API.")]
```

**Comment by mxinden** [comment]
Is `Rpc` used outside of the communication between the `NetworkBehaviour` and the `ConnectionHandler` implementation? In other words, is `Rpc` even used by any user code? If not, I suggest we don't treat removal of `Rpc` as a breaking change.

**Comment by thomaseizinger** [comment]
> Is `Rpc` used outside of the communication between the `NetworkBehaviour` and the `ConnectionHandler` implementation? In other words, is `Rpc` even used by any user code? If not, I suggest we don't treat removal of `Rpc` as a breaking change.

It is not but we don't necessarily know that :)

Is there much harm in deprecating it first?

**Review by jxs — APPROVED** [review]
Thanks Thomas!

**Comment by mxinden** [comment]
> > Is `Rpc` used outside of the communication between the `NetworkBehaviour` and the `ConnectionHandler` implementation? In other words, is `Rpc` even used by any user code? If not, I suggest we don't treat removal of `Rpc` as a breaking change.
> 
> It is not but we don't necessarily know that :)
> 
> Is there much harm in deprecating it first?

No harm in deprecating first. Just trying to find consensus for when we make a "breaking" change to it.

**Comment by thomaseizinger** [comment]
The rule of thumb I often apply is: If somebody does depend on it, can they fix it easily? Activating a feature that used to be default? Yes. Access a now private type, no.

**Pull request merged**